src/main/groovy/com/blackducksoftware/integration/hub/detect/bomtool/npm/NpmNodeModulesDependencyFinder.groovy

```diff
@@ -0,0 +1,154 @@
+import com.blackducksoftware.integration.hub.detect.util.FileFinder
+import com.google.gson.Gson
+import com.google.gson.stream.JsonReader
+
+@Component
+class NpmNodeModulesDependencyFinder {
+
+	private final Logger logger = LoggerFactory.getLogger(NpmNodeModulesDependencyFinder.class)
+
+	@Autowired
+	Gson gson
+
+	@Autowired
+	FileFinder fileFinder
+
+	/*
+	 * arraylist used to check for circular dependencies
+	 */
+	def modules = new HashSet<String>()
+
+	public DependencyNode generateDependencyNode(String nodeProjectDirectory) {
+		def packageJsonFile = fileFinder.findFile(nodeProjectDirectory, NpmConstants.PACKAGE_JSON)
+		def nodeModulesFile = fileFinder.findFile(nodeProjectDirectory, NpmConstants.NODE_MODULES)
+
+		PackageNode pNode
+
+		if(packageJsonFile?.exists()) {
+			pNode = deserializeJsonToPackageNode(packageJsonFile)
+		}
+
+		if(nodeModulesFile.exists() && nodeModulesFile.isDirectory()) {
+			if(pNode) {
+				logger.info("Traversing project directory starting from package.json dependencies")
+				return createNodeFromPackageNode(pNode, nodeModulesFile)
+			} else {
+				logger.info("No package.json found, parsing node_modules folder")
+				return createNodeFromNodeModules(nodeModulesFile)
+			}
+		} else if(pNode) {
+			/*
+			 * We have only the package.json and no node_modules folder which would require REST calls to complete.
+			 * This would be a good place to check for shrinkwrap/package-lock files
+			 */
+			logger.error("No node_modules folder located in target path. Can't run program")
+		}
+	}
+
+	/*
+	 * The best case scenario for parsing npm without npm installed (uses package.json and node_modules).
+	 * Note: Looks to be faster than the Cli finder
+	 */
+	public DependencyNode createNodeFromPackageNode(PackageNode pNode, File nodeModulesFile) {
+		DependencyNode node = createDependencyNode(pNode)
+
+		//Had to create a HashSet object before adding content because of issues with the keySet method for maps
+		def allDependencies = new HashSet<String>()
+		allDependencies.addAll(pNode?.dependencies?.keySet())
+		Set<String> optionalDependencies = pNode?.optionalDependencies?.keySet()
+
+		if(optionalDependencies) {
+			allDependencies.addAll(optionalDependencies)
+		}
+
+		for(String filename : allDependencies) {
+			if(modules.contains(filename)) {
+				continue
+			} else {
+				modules.add(filename)
+				def packageFile = fileFinder.findFile(nodeModulesFile.getAbsolutePath() + '/' + filename + '/', NpmConstants.PACKAGE_JSON)
+				if(packageFile) {
+					node.children.add(createNodeFromPackageNode(deserializeJsonToPackageNode(packageFile), nodeModulesFile))
+				}
+			}
+		}
+
+		node
+	}
+
+	/*
+	 * A slower check through the node_modules directory that doesn't require a starting package.json for the tree.
+	 */
+	public DependencyNode createNodeFromNodeModules(File nodeModulesFile) {
+		def pNode = new PackageNode();
+		pNode.name = 'BrianMandel'
```

Comment:
Why are we always setting the name to 'BrianMandel' and the version to the current day?